packages/@aws-cdk/cdk/lib/physical-name-generator.ts

```diff
@@ -0,0 +1,81 @@
+import crypto = require('crypto');
+import { IResource } from "./resource";
+import { Stack } from './stack';
+
+export class PhysicalNameGenerator {
+  public static from(resource: IResource): PhysicalNameGenerator {
+    return new PhysicalNameGenerator(
+        new PrefixNamePart(resource.node.stack.name, 25),
+        new SuffixNamePart(resource.node.uniqueId, 24),
+        resource.node.stack);
+  }
+
+  private readonly stackPart: NamePart;
+  private readonly idPart: NamePart;
+  private readonly stack: Stack;
+  private readonly hashLength: number;
+  private changeToLowerCase = false;
+
+  private constructor(stackPart: NamePart, idPart: NamePart, stack: Stack) {
+    this.stackPart = stackPart;
+    this.idPart = idPart;
+    this.stack = stack;
+    this.hashLength = 12;
+  }
+
+  public generate(): string {
+    const sha256 = crypto.createHash('sha256')
+      .update(this.stackPart.bareStr)
+      .update(this.idPart.bareStr)
+      .update(this.stack.env.region || '')
+      .update(this.stack.env.account || '');
+
+    const hash = sha256.digest('hex').slice(0, this.hashLength);
+
+    const parts = [this.stackPart, this.idPart]
+      .map(part => part.generate());
+
+    const ret = [...parts, hash]
+      .filter(part => part.length > 0)
+      .join('-');
```

Comment:
How do we know that all resource names can accept `-`. I rather we don't rely on this.

Reply:
If a resource doesn't, we can always override the generator in its L2.

Reply:
Let's keep it simple and remove the "-". I smell that they could be problematic.

Reply:
Removed.